streams/src/test/java/org/apache/kafka/streams/integration/KStreamRepartitionIntegrationTest.java

```diff
@@ -0,0 +1,859 @@
+    private List<KafkaStreams> kafkaStreamsInstances;
+
+    @Before
+    public void before() throws InterruptedException {
+        CLUSTER.createTopic(inputTopic, 4, 1);
+        CLUSTER.createTopic(outputTopic, 1, 1);
+
+        streamsConfiguration = new Properties();
+        kafkaStreamsInstances = new ArrayList<>();
+
+        streamsConfiguration.put(StreamsConfig.APPLICATION_ID_CONFIG, applicationId);
+        streamsConfiguration.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
+        streamsConfiguration.put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath());
+        streamsConfiguration.put(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG, 0);
+        streamsConfiguration.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, 100);
+        streamsConfiguration.put(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, Serdes.Integer().getClass());
+        streamsConfiguration.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serdes.String().getClass());
+        streamsConfiguration.put(StreamsConfig.TOPOLOGY_OPTIMIZATION, StreamsConfig.OPTIMIZE);
+    }
+
+    @After
+    public void whenShuttingDown() throws IOException {
+        kafkaStreamsInstances.stream()
+                             .filter(Objects::nonNull)
+                             .forEach(KafkaStreams::close);
+
+        IntegrationTestUtils.purgeLocalStreamsState(streamsConfiguration);
+        TEST_NUM.incrementAndGet();
+    }
+
+    @Test
+    public void shouldThrowAnExceptionWhenNumberOfPartitionsOfRepartitionOperationsDoNotMatchWhenJoining() throws InterruptedException {
+        final String topicB = "topic-b-" + TEST_NUM.get();
+        final String topicBRepartitionedName = "topic-b-scale-up";
+        final String inputTopicRepartitionedName = "input-topic-scale-up";
+        final int topicBNumberOfPartitions = 2;
+        final int inputTopicNumberOfPartitions = 4;
+        final AtomicReference<Throwable> throwable = new AtomicReference<>();
+
+        CLUSTER.createTopic(topicB, 1, 1);
+
+        final StreamsBuilder builder = new StreamsBuilder();
+
+        final Repartitioned<Integer, String> inputTopicRepartitioned = Repartitioned
+            .<Integer, String>as(inputTopicRepartitionedName)
+            .withNumberOfPartitions(inputTopicNumberOfPartitions);
+
+        final Repartitioned<Integer, String> topicBRepartitioned = Repartitioned
+            .<Integer, String>as(topicBRepartitionedName)
+            .withNumberOfPartitions(topicBNumberOfPartitions);
+
+        final KStream<Integer, String> topicBStream = builder
+            .stream(topicB, Consumed.with(Serdes.Integer(), Serdes.String()))
+            .repartition(topicBRepartitioned);
+
+        builder.stream(inputTopic, Consumed.with(Serdes.Integer(), Serdes.String()))
+               .repartition(inputTopicRepartitioned)
+               .join(topicBStream, (value1, value2) -> value2, JoinWindows.of(Duration.ofSeconds(10)))
```

Comment:
For this particular test, it seems we could detect the issue during topology `build()` already? Ie, we could do an additional early check? If we think it's worth doing, we should do it in follow up PR to not drag this PR any longer. \cc @vvcephei (If yes, we could change this test from an integration test to a unit test)

Reply:
I'll create followup ticket on that.

Reply:
Yeah, I'd agree with checking as early as possible in the special cases where we can know the partition counts statically. But also agree with doing it in a follow-on ticket, since it's kind of a nice-to-have.

Reply:
Cool. @lkokhreidze did you create a ticket already? (Just want to make sure we don't drop this on the floor.)

Reply:
@mjsax yes, here it is https://issues.apache.org/jira/browse/KAFKA-9850

Reply:
Thank you!